### ISSUE TITLE
[BUG] System::Windows::Forms::OpenFileDialog::AutoUpgradeEnabled does not work correctly with NetCore 8.0

### ISSUE THREAD

**Issue by chacha21** [dotnet/winforms#10846]
When setting AutoUpgradeEnabled to false in an System::Windows::Forms::OpenFileDialog

- everything goes right in a .Net 4.8 project
- no filenames are returned when exiting the dialog in a .Net 8.0 project (netcore8.0-windows)

It seems related to Multiselect is true

To be able to use System::Windows::Forms::OpenFileDialog in a NetCore 8.0 project, `<UseWindowsForms>true</UseWindowsForms>` must be used.

I don't know if it's relevant, but I have also referenced Microsoft.Windows.Compatibility in my project.

Is there any workaround ?

I need this feature to use customized dialogs.

My reproducer :

```
System::Windows::Forms::OpenFileDialog^ openFileDialog = gcnew System::Windows::Forms::OpenFileDialog();
openFileDialog->AutoUpgradeEnabled = false;
openFileDialog->Title = "Load...";
openFileDialog->AddExtension = true;
openFileDialog->Multiselect = true;
openFileDialog->DefaultExt = "cfg";
System::String^ filterString = "";
filterString += System::String::Format("{0}{1}",
  System::String::IsNullOrEmpty(filterString) ? "" : "|",
  "Cfg files" + " (*.cfg)|*.cfg");
filterString += System::String::Format("{0}{1}",
  System::String::IsNullOrEmpty(filterString) ? "" : "|",
  "All files" + " (*.*)|*.*");
openFileDialog->Filter = filterString;
System::Windows::Forms::DialogResult code = openFileDialog->ShowDialog(this);
//select one or a few files
System::Console::WriteLine("selected count = {0}", openFileDialog->FileNames->Length);//returns 0
```

**Comment by Dave-Lowndes** [comment]
Simple C# repo in a .NET8 console application project (with `<UseWindowsForms>true</UseWindowsForms>`):

```
using System.Windows.Forms;

namespace ConsoleApp1
{
    internal class Program
    {
        [STAThread]
        static void Main(string[] args)
        {
            var ofd = new System.Windows.Forms.OpenFileDialog();
            ofd.AutoUpgradeEnabled = false;
            ofd.Multiselect = true;
            if ( ofd.ShowDialog() == DialogResult.OK )
            {
                // Despite selecting a file, ofd.FileName & ofd.FileNames have nothing
                Console.WriteLine("Hello, World!" + ofd.FileNames[0]);
            }
        }
    }
}

```

**Comment by merriemcgaw** [comment]
@Olina-Zhang did we have this code working in a .NET 7 or earlier in .NET Core to see if we ever had it working or if it was a regression? Thanks!

**Comment by chacha21** [comment]
No chance to see that fixed before .NET 9.0 ? This is a real blocker for me.

**Comment by Dave-Lowndes** [comment]
@merriemcgaw It's OK when built for .NET 6 or 7 - problem only shows with .NET 8

**Comment by elachlan** [comment]
Reproduced in main. There is two off by one bugs. Submitted a PR at #10849

**Comment by MandiMan** [comment]
@merriemcgaw  verified this issue in .NET 5.0 & 6.0 & 7.0, when AutoUpgradeEnabled=false and Multiselect = true, No exceptions thrown, but in 8.0 & 9.0, throw an exception. 
5.0 & 6.0 & 7.0:
![Untitled](https://github.com/dotnet/winforms/assets/108860782/1cb4db68-bb70-47e3-904d-c1af8599c69c)
8.0 & 9.0:
![image](https://github.com/dotnet/winforms/assets/108860782/4d1cc7ad-919c-4c2a-9440-c778f836c0b3)

**Comment by lonitra** [comment]
> No chance to see that fixed before .NET 9.0 ? This is a real blocker for me.

The fix will be available in April .NET 8 servicing release.

**Comment by Nora-Zhou01** [comment]
Tested this issue in the latest .Net 9.0 SDK build: 9.0.100-preview.2.24116.21, it was fixed that No exception thrown when AutoUpgradeEnabled = false and Multiselect = true.
![10846](https://github.com/dotnet/winforms/assets/104609169/befd712c-ebc5-40db-9950-99b0285609ce)

**Comment by Dave-Lowndes** [comment]
@Nora-Zhou01 Did you check  for both a single file and multiple file selections?

**Comment by Nora-Zhou01** [comment]
No exception is thrown when selecting single and multiple files respectively.

https://github.com/dotnet/winforms/assets/104609169/28823edc-9eb9-4a86-87aa-3ea2edcf9130

**Comment by MelonWang1** [comment]
Verified this issue with .NET 9 Preview 2 test pass build, it was fixed. Test result is same as above.

**Comment by MelonWang1** [comment]
Verified issue on .NET 8.0.300 with dlls built from Release/8.0 branch of Winforms repo, the issue was fixed, test result is same as above.

**Comment by MandiMan** [comment]
Verified with .NET SDK 8.0.4 test pass build, this issue was fixed. Test result is same as above.

**Comment by PeterCunningham** [comment]
Many thanks for the rapid fix and its deployment to Visual Studio 17.9.6. A user encountered the problem in our application, Serafim FUTURES, so it is good to be able to solve the problem without needing a work-around and associated testing etc.
